### DISCUSSION TITLE
Control Points for SRT

### DISCUSSION THREAD

**Discussion by whugemann** [ImageMagick/ImageMagick#6878] [Help] [answered]
### ImageMagick version

7.1.1-15

### Operating system, version and so on

Windows 10

### Description

In contrast to '-distort affine' and '-distort perspective', you cannot use control points in the case of '-distort SRT', which is a shortcoming. (I know that I will get an SRT if I supply only two control points for either '-distort perspective' or '-distort affine'.)

But with the lack of '-distort SRT point1, point2, ...' you also lack the possibilty to supply more points in order to arrive at a least square solution as for '-distort affine' and '-distort perspective'.

In order to superimpose photographs, I usally use Fiji's SIFT plugin, where you can choose SRT as the desired transformation (called 'similarity' in this plugin), but sometimes I need to supply the control points manually, because SIFT fails to find suitable ones. In these cases, I would like to switch to ImageMagick as a fallback.

Could this option be added to ImageMagick -- or am I overlooking something?

**Comment by snibgo** [comment]
Isn't that what `-distort RigidAffine point1,point2 ...` does?

**Reply by whugemann** [reply, marked as answer]
Thanks for the hint, you are right. I didn't know about RigidAffine. Is that something new in Version 7? 
The documentation says:
"A rigid affine (also known as a Euclidean transform), is similar to Affine but restricts the distortion to 4 arguments (S, R, Tx, Ty) with Sy = Sx and Ry = -Rx so that the distortion only has scale, rotation and translation. No skew. A minimum of two control point pairs is required."

So, basically, it does the same as '-distort SRT', based on control points.
Which leaves us with the question why there are two different names for the same type of distortion ...

**Reply by fmw42** [reply]
Two functions because 1) the coding would be much harder in -distor SRT 2) the coding was an easy change to -distort Affine 3) it is similar to the name in other tools, 4) it is affine, but restricted so no skew flexibility, i.e. that means it is rigid.  Yes, this is new in IM 7, but was introduced quite a while back. It is not in IM 6 to my understanding.

**Reply by whugemann** [reply]
I still heavily rely on Anthony's examples for IM 6, where I didn't find an example for RigidAffine. It's a pity that these pages are not maintained anymore.
I am somewhat sceptical about the validity of argument one from a user's point of view.

**Reply by fmw42** [reply]
It was much easier to code in -distort Affine, since there was already code for control points. All that was needed was to restrict the two scaling arguments sx=sy when doing the least squares fit of the control points.  -distort SRT has no code for solving the least squares and was not set up in a way to do so.  So the functionality of -distort rigidaffine was a closer match to -distort Affine that it would be to -distort SRT.

**Reply by whugemann** [reply]
Fred, you are right from the programmer's point of view, but I bet that there would have been other solutions to this problem, hiding the internals from the user.
IMHO, the term 'rigid affine' ist not that easy to interpretate. I would associate 'rigid' with just translation + rotation. What exactly is 'rigid' when you scale something (beside angles).
Well, I guess, choices have already been made, so there is not much sense in me arguing ...

**Reply by fmw42** [reply]
See nomenclature for example:

https://docs.opencv.org/2.4/modules/video/doc/motion_analysis_and_object_tracking.html#estimaterigidtransform
https://nghiaho.com/?p=2208

also
https://en.wikipedia.org/wiki/Rigid_transformation


However, some nomenclature does associate rigid (euclidean) with only translation and rotation and when scale is involved it is called similarity.  Perhaps we should have done both. However, that would still have been done within the same framework as -distort Affine when control points are involved.

At this point, it is done and so I think you will have to just get used to the naming and way it was coded.
